### PR TITLE
MappedFieldType no longer requires equals/hashCode/clone

### PR DESCRIPTION
**Pull request by romseygeek**
With the removal of mapping types and the immutability of FieldTypeLookup in #58162, we no longer
have any cause to compare MappedFieldType instances.  This means that we can remove all equals
and hashCode implementations, and in addition we no longer need the clone implementations which
were required for equals/hashcode testing.  This greatly simplifies implementing new MappedFieldTypes,
which will be particularly useful for the runtime fields project.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:128]: This TODO doesn't make sense now that we don't extend lucene's FieldType

File: server/src/test/java/org/elasticsearch/index/mapper/BinaryFieldTypeTests.java
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/index/mapper/BinaryFieldTypeTests.java:29]: FieldTypeTestCase extensions that test only equals and hashCode can just be removed entirely.  BinaryFieldType doesn't have any search methods on it, and docvalues are tested in BinaryFieldMapperTests

File: server/src/test/java/org/elasticsearch/index/mapper/TextFieldTypeTests.java
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/index/mapper/TextFieldTypeTests.java:54]: We check that these modifications block merging in TextFieldMapperTests, so we can safely remove them here - this is just checking the `equals` implementation, which is now gone.

**Review by jtibshirani — COMMENTED** [review]
This is a nice simplification! I left one comment about an accidental change in a test.

I was also curious, do we have a standard practice as to when we implement `equals`/ `hashCode` versus omitting them? We certainly implement these on request and response objects, but I think we make a case-by-case choice for internal objects like `MappedFieldType` ?

**Comment by nik9000** [comment]
> I was also curious, do we have a standard practice as to when we implement `equals`/ `hashCode` versus omitting them? We certainly implement these on request and response objects, but I think we make a case-by-case choice for internal objects like `MappedFieldType` ?

I *think* we only do this one request and response objects so we can assert round tripping them.

File: test/framework/src/main/java/org/elasticsearch/index/mapper/FieldTypeTestCase.java
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/index/mapper/FieldTypeTestCase.java:28]: It might be worth dropping this in a followup. It is just about empty now.

**Review by nik9000 — COMMENTED** [review]
This is great! @jtibshirani found something worth fixing before merging, but it is wonderful to see so much subtracted code!

**Review by javanna — APPROVED** [review]
LGTM

File: modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/RankFeaturesFieldTypeTests.java
Comment (romseygeek) [reply modules/mapper-extras/src/test/java/org/elasticsearch/index/mapper/RankFeaturesFieldTypeTests.java]: Oh, good catch!  Autocomplete is not always my friend...

File: test/framework/src/main/java/org/elasticsearch/index/mapper/FieldTypeTestCase.java
Comment (romseygeek) [reply test/framework/src/main/java/org/elasticsearch/index/mapper/FieldTypeTestCase.java:28]: The mock shard contexts are still pretty widely used, so I think it's worth keeping for now.

**Review by jtibshirani — APPROVED** [review]

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Pull request merged**
